src/metabase/metabot/util.clj

```diff
@@ -41,47 +41,72 @@
-                                  (vec qp/around-middleware)
-                                  (fn [query _rff _context]
-                                    (add/add-alias-info
-                                     (#'qp/preprocess* query))))]
-                          (qp dataset_query nil nil))
-        {:keys [fields]} query]
-    (for [field fields
-          :let [[_ id m] field
-                alias (::add/desired-alias m)]]
-      [alias (alias-map id)])))
+(defn- add-desired-aliases
+  "Add the aliases generated by the query processor to each results metadata field."
+  [{:keys [dataset_query] :as model}]
+  (let [fields           (let [qp (qp.reducible/combine-middleware
+                                    (vec qp/around-middleware)
+                                    (fn [query _rff _context]
+                                      (add/add-alias-info
+                                        (#'qp/preprocess* query))))]
+                           (get-in (qp dataset_query nil nil) [:query :fields]))
+        field-ref->alias (reduce
+                           (fn [acc [f id m]]
+                             (let [alias (::add/desired-alias m)
+                                   x     (reduce (fn [acc [k v]]
+                                                   (cond-> acc
+                                                     (nil? (namespace k))
+                                                     (assoc k v)))
+                                                 {}
+                                                 m)]
+                               (cond-> acc
+                                 alias
+                                 (assoc
+                                   [f id (when (seq x) x)]
+                                   alias))))
+                           {}
+                           fields)]
+    (update model :result_metadata
+            (fn [result_metadata]
+              (map
+                (fn [{:keys [field_ref] :as rsmd}]
+                  (assoc rsmd :desired_alias (field-ref->alias field_ref)))
+                result_metadata)))))
 
-(defn inner-query
+(defn- add-inner-query
   "Produce a SELECT * over the parameterized model with columns aliased to normalized display names.
+  Add this result to the input model along with the generated column aliases.
   This can be used in a CTE such that an outer query can be called on this query."
-  [{:keys [id] :as model}]
-  (let [column-aliases (->> (aliases model)
-                            (map (partial apply format "\"%s\" AS %s"))
-                            (str/join ","))]
-    (mdb.query/format-sql
-     (format "SELECT %s FROM {{#%s}} AS INNER_QUERY" column-aliases id))))
+  [{:keys [id result_metadata] :as model}]
+  (let [column-aliases (or
+                         (some->> result_metadata
+                                  (map (comp
+                                         (fn [[column_name column_alias]]
```

Comment:
the fn signature here of `[column_name column_alias]` seems reversed from the `(juxt :desired_alias :sql_name)` projection. It's either confusing or wrong imo

Reply:
and i suspect it is wrong since some don't have a column name but do have an alias (and i suspect that means they are reversed and of course they all have a name but some don't have an alias?)

Reply:
It is confusing. `:desired_alias` is the value produced from qp. I'll change it around to be more clear.

Reply:
Renamed for clarity.